### PR TITLE
Polymorphic comments

### PR DESCRIPTION
**Pull request by miles-smith**
These changes overhaul the ability to comment on various resources with the intention being a more flexible solution for the future.

Notes:
- `ProjectComment` replaced with a polymorphic `Comment` model.
- Added a database migration to recreate existing `ProjectComment`s as `Comments`.
- Migrated comments and a reference to the original resource in its `metadata`.
- The `ProjectComment#comment_type` field has been migrated to the `tags` attribute on its respective `Comment` record.
- Side effect logic of creating `ProjectComment`s (e.g. setting of various approval fields) has been refactored out into specific routes/controllers/actions. This is a starter for 10 that solves some immediate problems and there's probably scope for DRYing it up at some point.
- It is hoped that the refactoring of the approval side effect logic might help with some of the test flakiness that has been observed for some time. No guarantees though!

### CODE REVIEW

**Comment by miles-smith** [comment]
I'll take a look at the Brakeman issues...

**Comment by miles-smith** [comment]
Aside - I'm now spotting some test flakiness with `test/integration/workflow_test.rb`. Anyone know if dodgy/broken markup (e.g. unclosed tags etc.) could be a factor?

File: test/models/ability_test.rb [hunk 1]
```diff
@@ -916,6 +916,28 @@ def setup
...
   end
 
+  test 'comments' do
+    user        = users(:standard_user1)
+    commentable = projects(:dummy_project)
+
+    comment_one = Comment.create!(
+      commentable: commentable,
+      user: user,
+      body: 'I like to move it, move it'
```
Comment (kenny-lee-1) [inline test/models/ability_test.rb:891]: :)

File: test/integration/add_comments_to_rejected_project_test.rb [hunk 2]
```diff
@@ -3,6 +3,8 @@
 
 class AddCommentsToRejectedProjectTest < ActionDispatch::IntegrationTest
   test 'normal user can not edit when project in approved state' do
+    skip # TODO: fix me?
```
Comment (kenny-lee-1) [inline test/integration/add_comments_to_rejected_project_test.rb]: do we need check this requirement with BA team?

**Comment by kenny-lee-1** [comment]
Neat rewiring of all the approval logic

File: test/fixtures/comments.yml [hunk 3]
```diff
@@ -2,3 +2,6 @@ test_comment:
   user: standard_user
   commentable: dummy_project (Project)
   body: Confused as a chameleon in a bag of Skittles.
```
Comment (kenny-lee-1) [inline test/fixtures/comments.yml:4]: Have we already lost these in the earlier migration?

File: test/integration/add_comments_to_rejected_project_test.rb (hunk 2 above)
Comment (miles-smith) [reply test/integration/add_comments_to_rejected_project_test.rb]: Good shout. I think this was a temporary `skip` that never got removed...

File: test/fixtures/comments.yml (hunk 3 above)
Comment (kenny-lee-1) [reply test/fixtures/comments.yml:4]: I've spotted it so it 's ok
`tags: [project_comment.comment_type]`

**Review by kenny-lee-1 — APPROVED** [review]

**Pull request merged**